parrot/src/pfs_dispatch.cc

```diff
@@ -1497,14 +1497,48 @@ static void decode_syscall( struct pfs_process *p, int entering )
 
 		case SYSCALL32_getgroups32:
 		case SYSCALL32_getgroups:
-			if (entering && pfs_fake_setgid)
-				divert_to_dummy(p,-EPERM);
+			if (entering && pfs_fake_setgid) {
+				/* Make sure the spare slot's open */
+				if (p->ngroups <= PFS_NGROUPS_MAX) {
+					for (int i = 0; i < p->ngroups; i++) {
+						if (p->groups[i] == p->egid)
+							return;
```

Comment:
Copy-paste error, my mistake. I had that part in a function before, and maybe it really should be in one.
My intent is that `setgroups` can be whatever, but on a call to `getgroups`, we add the egid if it's not present. If some program happens to frequently switch uid, call `getgroups`, and not drop extra groups, it might be possible to accumulate groups that way.
On Dec 8, 2015 9:27 PM, Patrick Donnelly <notifications@github.com> wrote:
>
> In parrot/src/pfs_dispatch.cc:
>
> > @@ -1497,14 +1497,48 @@ static void decode_syscall( struct pfs_process *p, int entering )
>
> >  
>
> >  		case SYSCALL32_getgroups32:
>
> >  		case SYSCALL32_getgroups:
>
> > -			if (entering && pfs_fake_setgid)
>
> > -				divert_to_dummy(p,-EPERM);
>
> > +			if (entering && pfs_fake_setgid) {
>
> > +				/* Make sure the spare slot's open */
>
> > +				if (p->ngroups <= PFS_NGROUPS_MAX) {
>
> > +					for (int i = 0; i < p->ngroups; i++) {
>
> > +						if (p->groups[i] == p->egid)
>
> > +							return;
>
> I think you meant to do break; instead of return; here but that's also probably not what you want. You want to jump out of the if on line 1502?
>
> Something things to think about: what should happen if we do setgid before getgroups? What happens if the setgroups omits the primary group id?
>
> —
> Reply to this email directly or view it on GitHub.